### ISSUE TITLE
Add support for early stopping and saving best model?

### ISSUE THREAD

**Issue by sandylaker** [open-mmlab/mmcv#403]
Hi,

currently the checkpoints are save at the end of every epoch, and only checkpoints of the nearest k epochs are kept.
Could you please add hooks for early stopping, and saving the top k model based on the validation metrics? This would be very helpful.

Thanks

**Comment by hellock** [comment]
Good suggestion.

**Comment by SilvioGiancola** [comment]
Hi! Is there any plan on implementing such feature? Thank you! :)

**Comment by zhouzaida** [comment]
> Hi! Is there any plan on implementing such feature? Thank you! :)

hi @SilvioGiancola , we have supported saving the best model. Please refer to https://github.com/open-mmlab/mmcv/blob/7540cf73ac7e5d1e14d0ffbd9b6759e83929ecfc/mmcv/runner/hooks/evaluation.py#L33.

**Comment by zhouzaida** [comment]
We are interested in supporting early stopping. Are you interested in implementing it?

**Comment by SilvioGiancola** [comment]
Thank you for your amazingly fast answer :) 
Do you also have a hook for early stopping? I would be interested in such a feature. If you have any idea/guidance on how to implement it, I don't mind giving a hand.

**Comment by zhouzaida** [comment]
Thanks for your contribution.
We don't have this hook yet. But you may add a new hook to implement it. We can monitor a metric and we stop training when the metric stop improving. The metric can be val_loss or val_acc and so on.  Here is an implementation of early stopping. https://github.com/Bjarten/early-stopping-pytorch/blob/master/pytorchtools.py

**Comment by zhouzaida** [comment]
> Thank you for your amazingly fast answer :)
> Do you also have a hook for early stopping? I would be interested in such a feature. If you have any idea/guidance on how to implement it, I don't mind giving a hand.

hi @SilvioGiancola , is there any progress?

**Comment by 24hours** [comment]
Hi, Im interested on this feature and I can implement it. 

I propose we implement `should_stop` flag in runner toggle by EarlyStoppingHook following the implementation from 
https://github.com/PyTorchLightning/pytorch-lightning/blob/3fcfd0214cc95fa7de49f6404421b5c6cd113c57/pytorch_lightning/callbacks/early_stopping.py#L204

I will make a pull request if its ok.

**Comment by zhouzaida** [comment]
> Hi, Im interested on this feature and I can implement it.
> 
> I propose we implement `should_stop` flag in runner toggle by EarlyStoppingHook following the implementation from https://github.com/PyTorchLightning/pytorch-lightning/blob/3fcfd0214cc95fa7de49f6404421b5c6cd113c57/pytorch_lightning/callbacks/early_stopping.py#L204
> 
> I will make a pull request if its ok.

You are very welcome to create a PR to implement it. In fact, the key is how to determine whether it is time to stop training. Of course, we can implement `should_stop` flag in the runner.
